api/service/prometheus/service.go

```diff
@@ -29,32 +51,83 @@ type Handler struct {
 }
 
 var (
-	svc = &Service{}
+	registryOnce sync.Once
+	svc          = &Service{}
+	config       Config
 )
 
+func getJobName(config Config) string {
+	var node string
+
+	// legacy nodes are harmony nodes: s0,s1,s2,s3
+	if config.legacy {
+		node = "s"
+	} else {
+		if config.nodetype == "validator" {
+			// regular validator nodes are: v0,v1,v2,v3
+			node = "v"
+		} else {
+			// explorer nodes are: e0,e1,e2,e3
+			node = "e"
+		}
+	}
+
+	return fmt.Sprintf("%s/%s%d", config.network, node, config.shard)
+}
+
 // NewService sets up a new instance for a given address host:port.
 // An empty host will match with any IP so an address like ":19000" is perfectly acceptable.
-func NewService(config nodeconfig.PrometheusServerConfig, additionalHandlers ...Handler) {
-	if !config.HTTPEnabled {
+func NewService(additionalHandlers ...Handler) {
+	if !config.enabled {
 		utils.Logger().Info().Msg("Prometheus http server disabled...")
 		return
 	}
 
+	utils.Logger().Debug().Str("Config", config.String()).Msg("Prometheus")
+	if svc.registry == nil {
+		svc.registry = prometheus.NewRegistry()
+	}
+	handler := promhttp.InstrumentMetricHandler(
+		svc.registry,
+		promhttp.HandlerFor(svc.registry, promhttp.HandlerOpts{}),
+	)
+
 	mux := http.NewServeMux()
-	mux.Handle("/metrics", promhttp.Handler())
+	mux.Handle("/metrics", handler)
 	mux.HandleFunc("/goroutinez", svc.goroutinezHandler)
 
 	// Register additional handlers.
 	for _, h := range additionalHandlers {
 		mux.HandleFunc(h.Path, h.Handler)
 	}
 
-	utils.Logger().Debug().Int("port", config.HTTPPort).
-		Str("ip", config.HTTPIp).
+	utils.Logger().Debug().Int("port", config.port).
+		Str("ip", config.ip).
 		Msg("Starting Prometheus server")
-	endpoint := fmt.Sprintf("%s:%d", config.HTTPIp, config.HTTPPort)
+	endpoint := fmt.Sprintf("%s:%d", config.ip, config.port)
 	svc.server = &http.Server{Addr: endpoint, Handler: mux}
+	if config.enablePush {
```

Comment:
Do you think it would be better to move this part of function in `svc.Start`?

Reply:
sure, I can move.